homeassistant/components/fan/reproduce_state.py

```diff
@@ -73,18 +82,39 @@ async def _async_reproduce_state(
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
-            # Call services to adjust the attributes
-            if attr in state.attributes and not check_attr_equal(
-                state.attributes, cur_state.attributes, attr
-            ):
-                data = service_data.copy()
-                data[attr] = state.attributes[attr]
-                service_calls[service] = data
-
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
+                    data[attr] = state.attributes[attr]
+            service_calls[SERVICE_TURN_ON] = data
+        else:
+            # The simple attributes are copied directly.
+            for attr, service in SIMPLE_ATTRIBUTES.items():
+                # Call services to adjust the attributes
+                if attr in state.attributes and not check_attr_equal(
+                    state.attributes, cur_state.attributes, attr
+                ):
+                    data = service_data.copy()
+                    data[attr] = state.attributes[attr]
+                    service_calls[service] = data
+
+            # Speed and modes are only changed if they are non-None.
+            # That's because they are mutually exclusive, but we always
```

Comment:
I'm having trouble following what we're trying to say here. Can you perhaps think of a way to rephrase? I get what you mean by them being mutually exclusive, but i don't understand what that has to do with being non-none or what "we always store all of them" means and why that matters.

Reply:
Let me know if the way I wrote it now makes sense.